mountpoint-s3/src/prefetch/part_stream.rs

```diff
@@ -0,0 +1,266 @@
+use std::{fmt::Debug, ops::Range, sync::Arc};
+
+use bytes::Bytes;
+use futures::task::SpawnExt;
+use futures::{pin_mut, task::Spawn, StreamExt};
+use mountpoint_s3_client::{types::ETag, ObjectClient};
+use mountpoint_s3_crt::checksums::crc32c;
+use tracing::{debug_span, error, trace, Instrument};
+
+use crate::checksums::ChecksummedBytes;
+use crate::prefetch::part::Part;
+use crate::prefetch::part_queue::unbounded_part_queue;
+use crate::prefetch::task::RequestTask;
+use crate::store::PrefetchReadError;
+
+/// A generic interface to retrieve data from objects in a S3-like store.
+pub trait ObjectPartStream {
+    type ClientError: std::error::Error + Send + Sync + 'static;
+
+    /// Spawns a request to get the content of an object. The object data will be retrieved in fixed size
+    /// parts and can then be consumed using [RequestTask::read]. Callers need to specify a preferred
+    /// size for the parts, but implementations are allowed to ignore it.
+    fn spawn_get_object_request(
+        &self,
+        bucket: &str,
+        key: &str,
+        if_match: ETag,
+        range: RequestRange,
+        preferred_part_size: usize,
+    ) -> RequestTask<Self::ClientError>;
+}
+
+/// The range of a [ObjectPartStream::spawn_get_object_request] request.
+/// Includes the total size of the object.
+#[derive(Clone, Copy)]
+pub struct RequestRange {
+    object_size: usize,
+    offset: u64,
+    size: usize,
+}
+
+impl RequestRange {
+    pub fn new(object_size: usize, offset: u64, size: usize) -> Self {
+        let size = size.min(object_size.saturating_sub(offset as usize));
+        Self {
+            object_size,
+            offset,
+            size,
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.size
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.size == 0
+    }
+
+    pub fn object_size(&self) -> usize {
+        self.object_size
+    }
+
+    pub fn start(&self) -> u64 {
+        self.offset
+    }
+
+    pub fn end(&self) -> u64 {
+        self.offset + self.size as u64
+    }
+
+    pub fn trim_start(&self, start_offset: u64) -> Self {
```

Comment:
Is it intentional that `start_offset > self.end()` is valid and just silently becomes a zero-length range? (Similar thing for `trim_end`).

Reply:
Yes. Rustdoc added to clarify.